### ISSUE TITLE
Ed25519 key generation support

### ISSUE THREAD

**Issue by max-wittig** [paramiko/paramiko#1136]
Nearly every key, exept the Ed25519 key offers a `generate` function in paramiko, which is used to generate a new key pair.

Is there a reason why the `ed25519` keytype doesn't have this class method?

From the documentation:
http://docs.paramiko.org/en/2.4/api/keys.html

![image](https://user-images.githubusercontent.com/6639323/34103452-0bdc38d4-e3ed-11e7-99ef-2f8cabf4e794.png)

**Comment by dopaminereceptionist** [comment]
Also interested in this. Should be reasonably straight forward.

**Comment by johnnybubonic** [comment]
for reference:

here's a python 2 implementation: https://github.com/pts/py_ssh_keygen_ed25519

here's the overview of the new "openssh-key-v1" format that ssh_ed25519 **always** uses: https://github.com/openssh/openssh-portable/blob/master/PROTOCOL.key 
(which you should add support for for RSA and ECDSA keys as well, since openssh has announced that'll be the default at some time in the future anyways)

hope that helps

**Comment by johnnybubonic** [comment]
@bitprophet looks like this should be much more simplified since the cryptography module [has now implemented ed25519](https://github.com/pyca/cryptography/issues/3509).

**Comment by paprickar** [comment]
@bitprophet do you have a gut feeling when this can be implemented?

**Comment by bitprophet** [comment]
I actually saw this independently the other day when digging into an unrelated key issue - I had need to run basic key generation and reached for Ed25519 out of "this is the current 'good key type' to use in a  vacuum". And yup. It's missing this!

I see I added it to the feature milestone some time ago. I have no firm ETA for it but see https://bitprophet.org/projects/#roadmap - this one absolutely qualifies for my next set of feature-adds. Expect I'd hope to get to this after the Python 2 removal release (again per that roadmap).

**Comment by utkonos** [comment]
Could the cryptography `.generate()` just be wrapped/used directly using some of the code I wrote there? Unless I'm totally mistaken, this issue could be solved with a few lines of code and a few unit tests, no?

I can definitely write the code and PR to close this issue, I just want to make sure I'm on the right path before committing.

**Comment by utkonos** [comment]
BTW: there's no `bits` or `curve` selection needed for generating `ed25519`. The method would just be `.generate()` and boom, it's done.

**Comment by utkonos** [comment]
I've been running the _just do it_ code in production for 6 months with no issues. I looked even more closely at the `dir()` of the `cryptography` private key object and the `paramiko` key object. Basically `c_pkey._raw_private_bytes()` results in the private key raw bytes from the `cryptography` key. That data is exactly what needs to go in `Ed25519Key._signing_key`. There is literally nothing else. All the other steps are to use existing `paramiko` code. If the `Ed25519Key` class just had a `raw_key` parameter and didn't have the exeption, there wouldn't be much else to do. You don't even need to wait for a `generate()` method to pass policy decisions.

https://github.com/paramiko/paramiko/blob/bbb8d57e917f5a9fefd25a85f3d10cd0c1a52ef5/paramiko/ed25519key.py#L65-L66

Here is the `cryptography` data:
<img width="831" alt="image" src="https://user-images.githubusercontent.com/2775739/209605791-3dd4fb5b-ea71-491e-a784-ecdef33ffaae.png">


Here is the `paramiko` data using the code I wrote to Rube Goldberg it into the format the current `paramiko` code can use.
<img width="832" alt="image" src="https://user-images.githubusercontent.com/2775739/209605903-3bc4afaf-a302-4dc9-b22b-d4a7ca3418e5.png">

**Comment by dd-ssc** [comment]
Any updates on this ?

Incidentally, I was looking for a python package to generate `ed25519` keys and was going to choose `paramiko` (mostly as I'd already heard the name from Ansible, Boto, Fabric, etc.). I'll go ahead with @utkonos [code](https://github.com/paramiko/paramiko/issues/1136#issuecomment-1160771520) for the moment and look forward to turning that into a pure paramiko implementation at some stage in the future ✅

Thanks for the hard work everyone - much appreciated! 👍

**Comment by bskinn** [comment]
@dd-ssc -- bitprophet has been working extensively on a new architecture for #387 and #23, but I believe that's more for authentication workflows rather than key generation.

Check out the [3.2 CHANGELOG](https://www.paramiko.org/changelog.html#3.2.0) to see the (_experimental_) additions in that release. I know he's planning staged addition and rollout of these features through multiple releases, so auth is going to be the focus for some time yet.

Once that's solidified, then AFAIK he's going to turn to other fixes and features.

**Comment by evan0greenup** [comment]
> I've been running the _just do it_ code in production for 6 months with no issues. I looked even more closely at the `dir()` of the `cryptography` private key object and the `paramiko` key object. Basically `c_pkey._raw_private_bytes()` results in the private key raw bytes from the `cryptography` key. That data is exactly what needs to go in `Ed25519Key._signing_key`. There is literally nothing else. All the other steps are to use existing `paramiko` code. If the `Ed25519Key` class just had a `raw_key` parameter and didn't have the exeption, there wouldn't be much else to do. You don't even need to wait for a `generate()` method to pass policy decisions.
> 
> https://github.com/paramiko/paramiko/blob/bbb8d57e917f5a9fefd25a85f3d10cd0c1a52ef5/paramiko/ed25519key.py#L65-L66
> 
> Here is the `cryptography` data: <img alt="image" width="831" src="https://user-images.githubusercontent.com/2775739/209605791-3dd4fb5b-ea71-491e-a784-ecdef33ffaae.png">
> 
> Here is the `paramiko` data using the code I wrote to Rube Goldberg it into the format the current `paramiko` code can use. <img alt="image" width="832" src="https://user-images.githubusercontent.com/2775739/209605903-3bc4afaf-a302-4dc9-b22b-d4a7ca3418e5.png">

When using this method to generate `ed25519` key, the key file is not passphrase protected.

**Comment by utkonos** [comment]
> When using this method to generate `ed25519` key, the key file is not passphrase protected.

Yes, that's correct. If you have a suggestion for adding password protection, please go ahead.

On a side note: I think the probability of ed25519 key generation being supported by paramiko is zero, so please help where you are able to.
